x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java

```diff
@@ -347,6 +347,7 @@ private void checkUniqueOrders(Map<Integer, Set<String>> orderToRealmName) {
 
     private List<RealmConfig> buildRealmConfigs() {
         final Map<RealmConfig.RealmIdentifier, Settings> realmsSettings = RealmSettings.getRealmSettings(settings);
+        RealmSettings.verifyRealmNameToDomainNameAssociation(settings, realmsSettings.keySet());
```

Comment:
Why don't we use the list of `Realm`s built by `Realms#initRealms`? I think it could help simplify the logic, e.g. no specially handling for default file/native?
I guess one reason for this approach is that `domain` is a field under `RealmConfig` so it must be computed before realms are intialised. But if we can let go `domain` being field in `RealmConfig` and just have the whole domain concept works standalone, it will probably produce code that is easier to follow and requires less special logic.

Reply:
We discussed offline. The `List<Realm>` does not contain information about disabled realms. Hence the validation will not be complete with it.